### ISSUE TITLE
Emissive Issues

### ISSUE THREAD

**Issue by OldFlak** [TheGameCreators/GameGuruRepo#970]
Entities that use Emissive are always white when turning down gamma.

This big sections of center light should be cyan not white.
There is also no real glow to emissives as in previous builds - they are just bright.
![Emissive Issue](https://user-images.githubusercontent.com/12246318/111020236-7b70be80-8418-11eb-803f-6190475d7308.png)

System Specs:
None detected!

**Comment by rvanner** [comment]
@LeeBamberTGC from your Teamwork list, likely the same issue;

Models with emissive textures assigned on existing levels - emissive color is white, can't change it...

**Comment by LeeBamberTGC** [comment]
@michaeloldman Could you attach a small test model that has emissive texture as shown in the shot, I can then reproduce quickly and ensure a future build allows the previous behavior. Maybe also a shot of what it should look like (before and after) with the small custom model would be ideal.

**Comment by michaeloldman** [comment]
Hi Lee,
Couple models that show the issue attached.

To replicate:
Open NEW Map
Plonk a start marker down and put the two models in the level;
- For each model turn the Emissive Strength up to 10

In test-game change these setting:
Customize Sky:
Simultaed Sky = ON (currently only way to get dark scenes)
Lighting Color = black
Ambient Color = black

Post Processing:
Bloom = ON (0.500)
All other settings = OFF
Gamma = 1.200

Also note:
Toggle Auto Exposure On\Off
ON  = Emissive glow is great (similar to what we had in pre-repo version)
OFF = Emissive glow is flat

Expected Results (pre-repo)
Emissive area of lights should be the color that is on the illumination texture - not white.

Hope that helps.

Regards
Michael

**Comment by michaeloldman** [comment]
Email sent :)

**Comment by michaeloldman** [comment]
We used to be able to get much better emissives in earlier builds:
![MaxLighting_02](https://user-images.githubusercontent.com/12246318/111925982-0b8cc300-8aff-11eb-94fd-95b7faf179cd.png)

**Comment by michaeloldman** [comment]
"Models with emissive textures assigned on existing levels - emissive color is white, can't change it..."

We can change the color now by using the color picker in Properties, but that is not desirable unless the texture only has one color in it - many illumination textures have **multiple** colors.

**Comment by michaeloldman** [comment]
Here be pic in current build. I tinkered the settings for more reflection:
Here you can see the reflection from the big lights are cyanish, but the light itself is white.

![Emissive_Color_Test](https://user-images.githubusercontent.com/12246318/111927001-36791600-8b03-11eb-8948-2d2ab348b22d.png)

**Comment by michaeloldman** [comment]
Another earlier version showing correct color being used from the Illumination Texture of the model - the cyan **is** cyan - no adjustments to the color-picker in Properties.

![Racha_01](https://user-images.githubusercontent.com/12246318/111934491-9c6e9900-8b15-11eb-888b-bf1fd099dcf2.png)

**Comment by michaeloldman** [comment]
Build 2021.02.04 - Exhibits same issue :)

**Comment by LeeBamberTGC** [comment]
@michaeloldman Could you email me the model from the above shot, I can then try and reproduce here and see if this has been fixed recently.

**Comment by michaeloldman** [comment]
Hi Lee,
Models, script, and fpm via dropbox link below.
Hopefully I sent correct models 😊

https://www.dropbox.com/s/8nx8c7ubxr9nn6k/TheGizmo.zip?dl=0

Regards
Michael

**Comment by LeeBamberTGC** [comment]
I checked the Wicked shader and the exposure value is multiplying the contributions from the green and red channel of your emissive texture, so they all end up white. You can change the emissive texture to a pure blue however to ensure no other colors are included as the strength is increased. See shot:

![image](https://user-images.githubusercontent.com/35840583/117693064-885e2400-b1b5-11eb-9f82-d4dea0baa46a.png)

As you say, the other way is to adjust the materials emissve color for control over a non pure channel color. 

Also, if the strength is kept small, then the CYAN can be retained up to a point (before it becomes white).

Here is the shader code in Wicked:

if (surface.emissiveColor.a > 0 && g_xMaterial.uvset_emissiveMap >= 0)
	{
		const float2 UV_emissiveMap = g_xMaterial.uvset_emissiveMap == 0 ? input.uvsets.xy : input.uvsets.zw;
		float4 emissiveMap = texture_emissivemap.Sample(sampler_objectshader, UV_emissiveMap);
		emissiveMap.rgb = DEGAMMA(emissiveMap.rgb);
		surface.emissiveColor *= emissiveMap;
	}

Any suggestions from anyone on another way to get a final emissive for better result, please post here! Hope all this makes sense. Let me know if this clears things up, and we can switch this to an Enhancement request to see if we can have more emissive options over and above what Wicked's latest repo is providing.

**Comment by LeeBamberTGC** [comment]
In the next build, emissive works as before, without the clamping on the bloom that was applied to the latest Wicked repo.

**Comment by LeeBamberTGC** [comment]
![image](https://user-images.githubusercontent.com/35840583/117829105-5955ba00-b26a-11eb-9707-ebbc673dd79b.png)

**Comment by rvanner** [comment]
@michaeloldman Is this fixed for you now?

**Comment by michaeloldman** [comment]
@rvanner No emissive is nowhere near as good as in pre-repo versions of max

In Lee's screenie they are definitely glowing nicely but the emissive colors are wrong
- the glow is blue
- but the texture is white

In this pic (also shown above) the emissive is cyan and is only using the texture as color source.
![cyan emissive](https://user-images.githubusercontent.com/12246318/130354244-fd4c842f-884e-4abb-9dd5-435eb39ededf.png)

To get anything like that now, you have to change the emissive color using the color dialog in properties to a darkish blue to try to get close to the original cyan color. That of course is fine if your texture only has one color in the emissive texture.

Most emissive textures have multiple colors.

**Comment by rvanner** [comment]
@LeeBamberTGC When you have time, can you comment on this issue please?

**Comment by michaeloldman** [comment]
@rvanner was in a rush to get out the door when posting my previous reply - got the pics muddled a bit. Have updated previous post to hopefully explain better :)

**Comment by OldFlak** [comment]
When MAX can do this then it is fixed - till then emissive is rubbish :)

https://sketchfab.com/3d-models/naelurec-portal-06ed295524a64ca0972a1f0aec0ee57d (edited)

Note this is only using the emissive texture to get the color - as is should.
In MAX if you make it glow that much the color will just be white.

Pre-repo got close - since then emissive is only good if you don't want a decent glow happening, or you are happy with white.

**Comment by LeeBamberTGC** [comment]
@OldFlak Okay, spent some time on this one (got carried away). It seems the differences are from two sources. First the latest Wicked engine applies somethingh called 'ACESFitted' to the post processing step which flattens and wipes out the vivid color of the original emissive color texture, and then second is that the MAX/Wicked bloom is nowhere near as aggressive as the SketchFab renderer, choosing a more subtle bloom effect.

That said, here is my collection of notes to attempt a resolution on this issue:

![image](https://user-images.githubusercontent.com/35840583/145872187-6a37c461-3c6e-4ade-8f73-b79a188274f8.png)

As you can see, on the right is a snapshot of the SketchFab render using an emissive source color of 121,255,255. As you can see the SketchFab post process waps up the final color to 192,255,255 and adds a sned load of color bleed (bloom), marked one in the shot. The square marked as two shows the actual MAX bloom by comparison. The backdrop image is of the new build for this week, which REMOVES the ACES conversion and sets the model emissive strength to just 2.5.  These two changes retains the final emissive color to match the SketchFab render, and should produce the desired result.

Pretty sure by removing ACES I will be upsetting someone else, but I am liking the new uber emissive feature we have for this weeks build, not least because when we ramp up the strength we can create stuff like this:

![image](https://user-images.githubusercontent.com/35840583/145872808-fa5f0005-08f4-4ef6-8909-5bcd927eb293.png)

Ouch!!  P.S. I changed the original emissive colors to match the SketchFab non-post process model for a better RGB comparison test :)

**Comment by OldFlak** [comment]
Hi Lee,
Thanks for looking at this again. I be you nemesis no doubt!

The glow is not the problem

  *   The glow even now is fine
  *   The problem is that the source of the glow (the bar) is always white.
So in the last image the glow is fine, but the color of the bar (the source) is more like white – this is the problem.
  *   What happens when you want a red bar or an orange one etc.

[Diagram, schematic  Description automatically generated]

Obviously the glow would need to be kept to a tolerance that works in the engine, but atm if you want the source of the glow to retain its color you have to have practically no glow at all.
Maybe it is one that Janos needs to look closer at as it used to work fine in pre-repo – but wicked update broke it, surely they have seen similar issue, unless they only ever try white lights?

  *   What about a blue, red, green, or yellow neon sign?

I will have a play when new build is up 😊

Thanks again for your awesome work.

Regards
Michael

**Comment by LeeBamberTGC** [comment]
@OldFlak You are essentially correct, you can have the color if you keep emissive to less than 3 (and 1 if you want the original color), but as you say this means there is almost no strength going into the bloom shader to create the glow. I did have a dig about to see if I could insert an artificial 'power' value to the bloom process but no luck so far.  The only trick you have is to keep to pure primary and secondary colors and keep one or two RGB channels at zero so there is nothing to multiply up, or settle for a subtle emissive then ramp up the bloom like crazy.  I think an improvement to the bloom power (and quality) will go a long way to improving this situation.  If anyone knows how to fiddle with the Wicked Engine bloom shader, let me know!

**Comment by OldFlak** [comment]
@LeeBamberTGC this is what I am talking about with the emissive.
This pick is from is from the Wicked site https://wickedengine.net/2017/08/30/voxel-based-global-illumination/
Notice how the source of the glow is the same color that is emanating from the object. 

![image](https://user-images.githubusercontent.com/12246318/147755670-138282d8-aa8a-4f26-b598-48451a2beb2d.png)

So in MAX we don't need to boost the glow (that already works great) we just need to stop the source becoming white. 
Wicked Engine can do it, as the picture from wicked site shows. :)

OldFlak....

**Comment by nomad5oul** [comment]
![image](https://user-images.githubusercontent.com/37256116/154088789-9c2dac0c-19cd-4cd5-9e0b-fde21fa7904b.png)

We need the lights in Max to look like this with illumination and emissive volumetric glow.

**Comment by nomad5oul** [comment]
![image](https://user-images.githubusercontent.com/37256116/154089718-a98301cf-4753-450c-9a3f-576327846b5e.png)

**Comment by LeeBamberTGC** [comment]
Tagged as an ENHANCEMENT for the EMISSIVE VOLUMETRICS - It might just need better BLOOM to achieve this. It would be good if the above assets can be sourced so we can import them and see if we can get close to the shots above.

**Comment by OldFlak** [comment]
@LeeBamberTGC
Yes the BLOOM slider has doesn't really do anything much as far as affecting glow around lights
- 1.333 is default
- if you go higher it cancels out bloom on lights
- if you go lower it affects is a little but does not really make any difference

**Comment by nomad5oul** [comment]
[https://www.youtube.com/watch?v=rjMj4N-kUjU](url)
here is an earlier video from Max which had the correct emissive and volumetric lighting

![image](https://user-images.githubusercontent.com/37256116/154538463-fe0020f4-41d6-43da-8ec7-90a429d76518.png)

**Comment by ArgentArts** [comment]
Late to the game here. And correct me if I am wrong, but what I am used to is emission working this way:

If the emissive texture is black and white, then black is non-emissive and white is emissive and when you set an emissive color, then the white areas will be colored by the emissive color you select (i.e. you can color the white areas via the emissive color slider). 

If the emissive texture is black and any other color, then the other colors DETERMINE the emissive color themselves, black being non-emissive. In this case, the emissive color slider would have NO EFFECT on anything that was NOT WHITE. Thus, if an emissive texture were black and blue and red, then the blue and red areas would glow blue and red. Any color set via the emissive color setting should, then, have zero effect. The emissive texture should override this.

You would then use the Emissive Strength slider to set how powerful the glow is.

**Comment by OldFlak** [comment]
Just to add feedback from 2022.03.04 latest build. 
Emissive is pretty much toast now:
https://youtu.be/XS6pxXt8LA0

**Comment by ArgentArts** [comment]
What is the emissive color of your emissive texture in these images?

**Comment by ArgentArts** [comment]
Yeah, I have to agree. It's like you have to find a complementary emissive color to get your emissive texture color to glow properly, and that should not be how it works at all. This makes things especially difficult if your emissive texture has multiple colors, like red and blue.

**Comment by LeeBamberTGC** [comment]
@ArgentArts The nearest I can get to this is that we need work on both emissive color choices AND bloom.  If we just magnify any non-secondary color, it will always become white.  Not ideal.  Instead ramping the bloom factor oif your chosen hue will be much better, Not sure if Wicked graphics engine supports this now, but we can look at it after EA.

**Comment by ArgentArts** [comment]
From what I recall in other engines, if your emissive textures is WHITE and BLACK, then setting an emissive color affects the WHITE (i.e. choosing a red color would make it the white areas red). However, if your emissive texture is black (non-emissive), red, and blue, then choosing an emissive color would have NO EFFECT on color (as far as adding any color). Your emissive texture colors would determine the colors in this case (blue and red). The emissive color chose would only effect intensity (or nothing at all, since there is an intensity slider). Is this incorrect? 

In other words, I had believed that as long as the emissive texture was non-white, it would override any emissive color choice. This way the end-user could, for example, set up a glowing control panel with a bunch of differently colored glowing controls (red, green, and yellow glowing lights and buttons, for example).

**Comment by OldFlak** [comment]
@LeeBamberTGC Hi Lee thanks for looking at this.
As ArgentArts mentions, and discussed throughout this thread, emissive texture should be used for color not the color picker unless the the emissive texture is black and white.

Fixing bloom would be great, it is practically useless atm and has always been that way. 
You literally have between 0 and 1.5 (2 at max) to have any effect on the range for emissive glow. 1.5 being glow (sort of) and zero being max glow possible.
- surely it should be from 0-100 - 0 being now bloom and 100 completely maxing it out?

But as the videos show above it doesn't matter what method you use you cannot get color like the image form Wicked site.
![image](https://user-images.githubusercontent.com/12246318/157764578-2b7a33b6-982f-4046-8657-f85856a6aaea.png)

**Comment by ArgentArts** [comment]
Any word on emissive being fixed?

**Comment by ArgentArts** [comment]
@LeeBamberTGC I'm concerned after listening to today's live broadcast about your conclusions for how to fix or improve emissive materials in MAX. It's not about selecting an emissive color. That's not the issue we're seeing here. As we've said previously, the emissive color should be selected by the actual emissive TEXTURE. So, if my emissive texture is black (no emissive value) and BLUE, then BLUE is the emissive color. That BLUE should OVERRIDE any other color, selected or otherwise. And if we crank up the emissive value (how powerful it is) then it should just glow a brighter and brighter blue of that shade, like in OldFlak's screen shot above. 

The only time selecting an emissive color should do anything at all is if the color on the emissive texture is WHITE.

**Comment by LeeBamberTGC** [comment]
@ArgentArts Yep that is what I meant, the texture provides the color and the strength ultimately decides how much of the scene is invaded by that precise color :)

**Comment by ArgentArts** [comment]
@LeeBamberTGC Awesome! It sounded like you were talking about the color selector choosing the color and I was like, nooooooo! (Luke Skywalker scream!) ;)

**Comment by OldFlak** [comment]
> the texture provides the color and the strength ultimately decides how much of the scene is invaded by that precise color :)

Yes - now we are on the same page @LeeBamberTGC :)

**Comment by nomad5oul** [comment]
This sounds like an excellent approach. We really need emissive to have that volumetric look on the lights which looks WOW

**Comment by OldFlak** [comment]
>This sounds like an excellent approach. 
Tis not just an excellent approach, it is the way Emissive should work :)

Not sure why has it taken such a long thread to get to this point - lol :) 
This is what has been repeatedly mentioned all through this thread :)
Starting here:
https://github.com/TheGameCreators/GameGuruRepo/issues/970#issuecomment-803689648

Lets hope we get it sorted soon :)

**Comment by LeeBamberTGC** [comment]
@PaulSJ I have assigned this to you, specifically to look at the bloom component of Wicked and see if there are settings that can increase the effect. Right now our bloom is pretty weak and cannot achieve the kind of bloom in the shots above.

**Comment by ArgentArts** [comment]
@LeeBamberTGC @PaulSJ It's not just the bloom. I can get a decent bloom at high settings. It's when you crank up the strength of the emission the core color turns white quickly. Too quickly.

**Comment by LeeBamberTGC** [comment]
@ArgentArts Can you send me a non-MAX screenshot of a scene with high strength emissive that has low bloom and a non primary/secondary color emitting, thanks!

The battle in my mind is that when you modulate a non primary ot secondary color, the maths always makes them white:

RGB(1,2,3) * 1000 = RGB(255,255,255)
vs 
RGB(1,0,0) * 1000 = RGB(255,0,0)

**Comment by nomad5oul** [comment]
![image](https://user-images.githubusercontent.com/37256116/165798264-71b6a679-d530-4270-b665-8567025eccc0.png)
![image](https://user-images.githubusercontent.com/37256116/165798313-2dfb1192-a051-4437-abac-1d0e68e62e93.png)

We just need the emissive / bloom combination to look like this

**Comment by nomad5oul** [comment]
Yes we need the volumetric aura around the emissive parts of a model. thats what gives it the WOW factor

**Comment by ArgentArts** [comment]
This one shows the emissive texture better because this one has two emissive colors on the one texture:

![emit4](https://user-images.githubusercontent.com/36735507/165801199-b3c0a237-4267-4d2b-a109-61eb4f01f599.png)

Again, the settings are very high to exaggerated the effect. The color is set to white so as not to effect the texture (i.e. let the texture color take control of the bloom and overall color of the emissive color effect).

**Comment by LeeBamberTGC** [comment]
@ArgentArts Can you show this same shot in MAX with BLOOM set to MAXIMUM, thanks!

**Comment by ArgentArts** [comment]
@LeeBamberTGC Here you go:

![emit5](https://user-images.githubusercontent.com/36735507/165810231-e19d7ae1-66b1-411f-a074-856e03400dbe.png)

Not the exact same scene, but the same props.

**Comment by LeeBamberTGC** [comment]
@ArgentArts So would you agree that when we crank up the emissive power, we clamp it at a max of 1 to retain the original texture color, and anything over 1.0 affects the bloom strength, this should yield something similar to the non-MAX shot.

**Comment by ArgentArts** [comment]
@LeeBamberTGC Personally (and others may disagree with me), I think it should be something in-between the two shots. We definitely need to be able to make emission look like the shots shown in the non-MAX screens I've shown. But I think we also need to be able to create the others as well (as I believe they are more realistic, just less ... wow ... and less artistic). So, I think we need both. As a result, I think we need to retain the original texture color of emission over 1.0 up a point. But there also needs to be a point where it starts to wash out the original color, too (maybe at the more extreme end?). Again, just my opinion.

**Comment by LeeBamberTGC** [comment]
@ArgentArts I think this is another of those "let the user choose" moments, so we don't need to decide one way over the other.  Perhaps we keep the emissive strength so you can modulate it up to 64.0, and then perhaps a second value for emisive which can override the bloom strength just for that texture color to give it more OOOMMPH!  @PaulSJ Do you agree?

**Comment by ArgentArts** [comment]
@LeeBamberTGC That actually sounds good to me. So, two sliders - one controls the strength of the bloom/emissive strength and the second to override (or not) the color of the texture. That could actually add some cool effects. ;)

**Comment by Granada1** [comment]
Sounds like a great compromise, good for everyone

**Comment by nomad5oul** [comment]
OldFlak will be very happy emissive is getting fixed / improved.
It will be another AAA feather in Max's visual cap which is fantastic

**Comment by ArgentArts** [comment]
@nomad5oul He's not the only one! ;)

**Comment by nomad5oul** [comment]
LOL. I basically want my Max games to look like a volumetric haze of lights, colors, fog and particles.

**Comment by OldFlak** [comment]
The eimissive glow is governed by both the emissive strength and the bloom slider
Can we have the bloom slider itself also work correctly? For me O.1 is about as much glow as it will allow, whereas 100 has no effect at all. 

0 should be bloom off 
100 should be bloom maxed out

Many of MAX sliders seem to work like that but it doesn't make sense to the user.

**Comment by PaulSJ** [comment]
So I've done a bit of research on this and from what I can tell the emissive color should desaturate as it gets brighter, but I agree that going full white is too much. WickedEngine seems to be following Unreal's changes which is why it used to treat emissive differently, found here [https://docs.unrealengine.com/5.0/en-US/color-grading-and-the-filmic-tonemapper-in-unreal-engine/](url). I've adjusted the tonemapping so it won't wash out colors so agressively, but it will still destaurate them as they get brighter.

The bloom threshold slider isn't really meant to control the strength, it determines the brightness an object needs to reach before it will contribute to bloom. So the higher it is the less bloom you will get because it rasies the brightness that objects need to reach before they bloom.

I've added a Bloom Strength slider that will determine how strongly the final bloom is applied to the scene. So you can really crank it up if you want without having to increase the emissive strength any further.

**Comment by MonkeyFrogStudio** [comment]
Awesome! And thank you.

When drawing or painting, I remember something I was once told, "There's what's real and there's what looks right or good." The point that the person was trying to make was, sometimes what's correct or right or proper isn't always appealing. And, let's face it, sometimes we want to create fantasy. This applies even to emissive textures. Sometimes the emissiveness, bloom, etc., should be just as it is in reality. But there are times the artist creating the scene will want to create what "looks right or good" or even what's just pure fantasy. We need to tools to do all of that.

**Comment by nomad5oul** [comment]
This sounds fantastic. Thankyou :)

**Comment by OldFlak** [comment]
@PaulSJ - Now that is cool, great work - thanks!

![image](https://user-images.githubusercontent.com/12246318/167121318-ecf5230c-e346-4ebb-a911-8a15bf46e29f.png)

If all the others that have chimed in on this thread are happy - I be OK to close this thread - finally :)
